### PR TITLE
Add listener module to Gdrive connector

### PR DESCRIPTION
**Pull request by SkNuwanTissera**
### Purpose

- This PR will add listener support for the Google drive connector.
- Documentation for the listener module also added to this PR.
- Add icon in display annotation.
- Updated documentation in both client and listener.
- Increase drive version by one.
- Included java wrapper implementation for the listener. 
- Added a scheduler for watch operations.
- Include migration changes to Alpha4.

### Goals

- Add listener module for Gdrive connector.
- Resolves https://github.com/wso2-enterprise/choreo/issues/2696
- Listener can be configured with a predefined value for expiration time for watch operations. 

### Approach

- User can listen for changes within a specific file, specific folder or whole drive. That is configurable on listener initiation.
- Google drive watch API used to watch files & folder events.
- Push notifications are used to trigger the listener with incoming changes.
- Watch expiration time configuration
Default : 3600s. If we are watching for a specific file or folder resource, the maximum expiration time is 86400s(1 day). If we are watching for all changes in the drive, the Maximum expiration time is 604800s(1 week). If the user has configured a value greater than the maximum expiration, we are setting it to the maximum expiration time.

### User stories
Almost all user stories are supported.
1. Listen to the new file creation event.
2. Listen to the new folder creation event.
3. Listen to the file update event.
4. Listen to the folder update event.
6. Listen to the permanently deleted event.
7. Listen to trash event.

### Release note
Listener support is added to Gdrive Connector.

### Documentation
https://github.com/SkNuwanTissera/module-ballerinax-googleapis.drive-1/blob/main/modules/listener/README.md

Training
N/A

Certification
N/A

Marketing
N/A

Automation tests
Unit tests
N/A

Integration tests
N/A

Security checks
Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
Ran FindSecurityBugs plugin and verified report? no
Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes
Samples
Samples added into the samples directory.

Related PRs
N/A

Migrations (if applicable)
N/A

Test environment
JDK 11, Swan Lake Alpha2

Learning
N/A

### CODE REVIEW

File: modules/listener/constants.bal [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+const string FOLDER = "application/vnd.google-apps.folder";
+const string FILE = "application/vnd.google-apps.file";
+const string SPREADSHEET = "application/vnd.google-apps.spreadsheet";
+const string DOCS = "application/vnd.google-apps.document";
+const string GOOGLE_DRAWING = "application/vnd.google-apps.drawing";
+const string FORM = "application/vnd.google-apps.form";
+const string PHOTO = "application/vnd.google-apps.photo";
+const string PRESENTATION = "application/vnd.google-apps.presentation";
+const string VIDEO = "application/vnd.google-apps.video";
+const string audio = "application/vnd.google-apps.audio";
```
Comment (abeykoon) [inline modules/listener/constants.bal:27]: should be in CAPS as it is a constant.

File: modules/listener/listener.bal [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+    # Listener initialization
+    public function init(ListenerConfiguration config) returns @tainted error? {
+        self.eventService = config.eventService;
+        self.httpListener = check new (config.port);
+        self.driveClient = check new (config.clientConfiguration);
+        if (config.specificFolderOrFileId is string) {
+            self.isAFolder = check checkMimeType(self.driveClient, config.specificFolderOrFileId.toString());
+        }
+        if (config.specificFolderOrFileId is string && self.isAFolder == true) {
```
Comment (abeykoon) [inline modules/listener/listener.bal:60]: What is meant by `specificFolderOrFileId `? Can it be empty also?

File: modules/listener/listener.bal [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    private boolean isWatchOnSpecificResource = false;
+    private boolean isAFolder = true;
+
+    # Listener initialization
+    public function init(ListenerConfiguration config) returns @tainted error? {
+        self.eventService = config.eventService;
+        self.httpListener = check new (config.port);
+        self.driveClient = check new (config.clientConfiguration);
+        if (config.specificFolderOrFileId is string) {
+            self.isAFolder = check checkMimeType(self.driveClient, config.specificFolderOrFileId.toString());
```
Comment (abeykoon) [inline modules/listener/listener.bal:58]: Shall we rename `isAFolder ` to `isFolder `?

File: modules/listener/listener.bal (hunk 2 above)
Comment (abeykoon) [reply modules/listener/listener.bal:60]: Can't we simply use self.isAFolder as it is a boolean field?

File: modules/listener/listener.bal [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        }
+        if (config.specificFolderOrFileId is string && self.isAFolder == true) {
+            check validateSpecificFolderExsistence(config.specificFolderOrFileId.toString(), self.driveClient);
+            self.specificFolderOrFileId = config.specificFolderOrFileId.toString();
+            self.watchResponse = check startWatch(config.callbackURL, self.driveClient, self.specificFolderOrFileId.
+            toString());
+            check getCurrentStatusOfDrive(self.driveClient, self.currentFileStatus, 
+            self.specificFolderOrFileId.toString());
+            self.isWatchOnSpecificResource = true;
+        } else if (config.specificFolderOrFileId is string && self.isAFolder == false) {
```
Comment (abeykoon) [inline modules/listener/listener.bal:68]: Same here. shall we use !self.isAFolder?

File: modules/listener/listener.bal [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    
+    public isolated function attach(http:Service s, string[]|string? name) returns error? {
+        return self.httpListener.attach(s, name);
+    }
+
+    public isolated function detach(http:Service s) returns error? {
+        return self.httpListener.detach(s);
+    }
+
+    public isolated function 'start() returns error? {
```
Comment (abeykoon) [inline modules/listener/listener.bal:93]: Are we defining our own method or should we override `start()` method Listener provides us? Please check on this.

File: modules/listener/listener.bal [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        return self.httpListener.'start();
+    }
+
+    public isolated function gracefulStop() returns error? {
+        return ();
```
Comment (abeykoon) [inline modules/listener/listener.bal:98]: Shouldn't we send a message to google API here stating that we are stopping? https://developers.google.com/calendar/v3/push#stopping-notifications

File: modules/listener/listener.bal [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+            drive:ChangesListResponse[] response = check getAllChangeList(self.currentToken, self.driveClient);
+            foreach drive:ChangesListResponse item in response {
+                self.currentToken = item?.newStartPageToken.toString();
+                if (self.isWatchOnSpecificResource && self.isAFolder) {
+                    log:print("Folder watch response processing");
+                    check mapEventForSpecificResource(<@untainted> self.specificFolderOrFileId, <@untainted> item, 
+                    <@untainted> self.driveClient, <@untainted> self.eventService, <@untainted> self.currentFileStatus);
+                    check getCurrentStatusOfDrive(self.driveClient, self.currentFileStatus, self.specificFolderOrFileId);
+                } else if (self.isWatchOnSpecificResource && self.isAFolder == false) {
+                    log:print("File watch response processing");
```
Comment (abeykoon) [inline modules/listener/listener.bal:128]: These should be DEBUG level logs. Otherwise there will be a log line per event when we use the connector.

Comment (abeykoon) [reply modules/listener/listener.bal:128]: Please check all logs for above.

File: modules/listener/listener.bal [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+                    self.currentFileStatus);
+                    check getCurrentStatusOfFile(self.driveClient, self.currentFileStatus, self.specificFolderOrFileId);
+                } else {
+                    log:print("Whole drive watch response processing");
+                    check mapEvents(item, self.driveClient, self.eventService, self.currentFileStatus);
+                    check getCurrentStatusOfDrive(self.driveClient, self.currentFileStatus);
+                }
+            }
+        }
+        log:print("< RECEIVED >");
```
Comment (abeykoon) [inline modules/listener/listener.bal:139]: Do we need log lines like this? What does this log means?

File: modules/listener/listener.bal (hunk 3 above)
Comment (SkNuwanTissera) [reply modules/listener/listener.bal]: Hi
Renamed to isFolder in all occurrences.
Thank you

File: modules/listener/listener.bal (hunk 4 above)
Comment (SkNuwanTissera) [reply modules/listener/listener.bal]: Hi
Renamed to isFolder in all occurrences.
Thank you

File: modules/listener/listener.bal (hunk 5 above)
Comment (SkNuwanTissera) [reply modules/listener/listener.bal]: Hi
We are overriding as follows
    public isolated function 'start() returns error? {
        check self.httpListener.'start();
    }

Thank you

File: modules/listener/listener.bal (hunk 6 above)
Comment (SkNuwanTissera) [reply modules/listener/listener.bal]: Hi 
Yeah.. Handled stopWatch() in stop methods.
`public isolated function detach(service object {} s) returns error? {
        check stopWatchChannel(self.driveClient, self.channelUuid, self.watchResourceId);
        log:printDebug("Unsubscribed from the watch channel ID : " + self.channelUuid);
        return self.httpListener.detach(s);
    }

    public isolated function gracefulStop() returns error? {
        check stopWatchChannel(self.driveClient, self.channelUuid, self.watchResourceId);
        log:printDebug("Unsubscribed from the watch channel ID : " + self.channelUuid);
        return self.httpListener.gracefulStop();
    }

    public isolated function immediateStop() returns error? {
        check stopWatchChannel(self.driveClient, self.channelUuid, self.watchResourceId);
        log:printDebug("Unsubscribed from the watch channel ID : " + self.channelUuid);
        return self.httpListener.immediateStop();
    }`
Thanks

File: modules/listener/listener.bal (hunk 7 above)
Comment (SkNuwanTissera) [reply modules/listener/listener.bal]: Hi 
Changed to suitable log levels  in all occurrences
Thanks

File: modules/listener/listener.bal (hunk 8 above)
Comment (SkNuwanTissera) [reply modules/listener/listener.bal]: Removed the unwanted log lines in all occurences

File: modules/listener/constants.bal (hunk 1 above)
Comment (SkNuwanTissera) [reply modules/listener/constants.bal]: Changed to full block.
Thanks

**Comment by SkNuwanTissera** [comment]
New PR was initiated after squashing all the commits.
https://github.com/ballerina-platform/module-ballerinax-googleapis.drive/pull/3 
Hence, Closing this PR.

**Pull request closed without merging**